components/board/genericlinux/pin_types.go

```diff
@@ -41,3 +33,61 @@ type NoBoardFoundError struct {
 func (err NoBoardFoundError) Error() string {
 	return fmt.Sprintf("could not determine %q model", err.modelName)
 }
+
+// PinDefinition describes a gpio pin on a linux board.
+type PinDefinition struct {
+	Name            string `json:"name"`
+	Ngpio           int    `json:"ngpio"` // this is the ngpio number of the chip the pin is attached to
+	LineNumber      int    `json:"line_number"`
+	PwmChipSysfsDir string `json:"pwm_chip_sysfs_dir,omitempty"`
+	PwmID           int    `json:"pwm_id,omitempty"`
+}
+
+// PinDefinitions describes a list of pins on a linux board.
+type PinDefinitions struct {
+	Pins []PinDefinition `json:"pins"`
+}
+
+// UnmarshalJSON handles setting defaults for pin configs.
+func (conf *PinDefinition) UnmarshalJSON(text []byte) error {
+	type TempPin PinDefinition // needed to prevent infinite recursive calls to UnmarshalJSON
+	aux := TempPin{
+		Ngpio:      -1,
+		LineNumber: -1,
+		PwmID:      -1,
+	}
+	if err := json.Unmarshal(text, &aux); err != nil {
+		return err
+	}
+	*conf = PinDefinition(aux)
+	return nil
+}
+
+// Validate ensures all parts of the config are valid.
+func (conf *PinDefinition) Validate(path string) error {
+	if conf.Name == "" {
+		return utils.NewConfigValidationFieldRequiredError(path, "name")
+	}
+
+	if conf.Ngpio == -1 {
+		return utils.NewConfigValidationFieldRequiredError(path, "ngpio")
+	}
+
+	if conf.LineNumber == -1 {
+		return utils.NewConfigValidationFieldRequiredError(path, "line_number")
+	}
+
+	if conf.LineNumber < 0 {
+		return utils.NewConfigValidationError(path, errors.New("line_number on gpio chip must be greater than zero"))
+	}
+
+	if conf.LineNumber >= conf.Ngpio {
+		return utils.NewConfigValidationError(path, errors.New("line_number on gpio chip must be less than ngpio"))
+	}
+
+	if conf.PwmChipSysfsDir != "" && conf.PwmID == -1 {
+		return utils.NewConfigValidationError(path, errors.New("must supply pwm_id for the pwm chip"))
```

Comment:
What about the opposite condition, where the PWM ID is set but the directory is not? I suspect we need another check for that.

Reply:
we already had this discussion in a different pr- one option we had was to log a warning instead, but didn't do this cuz Rand did not like the logging happening in setup